src/main/java/CRDT.java

```diff
@@ -1,21 +1,67 @@
-import java.io.Serializable;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.List;
 
-import com.ericsson.otp.erlang.OtpErlangBinary;
+import java.lang.reflect.Method;
 
-interface CRDT extends Serializable {
-    void invoke(GenericFunction obj);
+abstract class CRDT implements Antidote_interface {
+  private static final long serialVersionUID = 1L;
 
-    Object read();
+  abstract public Object value();
 
-    void snapshot();
+  public void invoke(GenericFunction obj) {
+    try {
+      String method_name = obj.getFunctionName();
+      List<Object> args = obj.getArguments();
+      /* Integer id = obj.getId(); */
 
-    // just serialize for snapshot read
+      Class<?>[] argTypes = (Class[]) this.getClass().getField(method_name).get(this);
 
+      Method method;
+      try {
+        method = this.getClass().getDeclaredMethod(method_name, argTypes);
+
+        method.invoke(this, args.toArray());
+      } catch (NoSuchMethodException e) {
+        // This is the case where the method has generic args
+        // Still doesn't work great
+        for (int i = 0; i < args.size(); i++) {
+          argTypes[i] = Object.class;
+        }
+
+        method = this.getClass().getDeclaredMethod(method_name, argTypes);
+        method.invoke(this, args.toArray());
+      }
+
+    } catch (Exception e) {
+      e.printStackTrace();
+      throw new RuntimeException(e);
+    }
+  }
+
+  public CRDT deepClone() {
     /*
-     * public static int bin_to_int(OtpErlangBinary val) { return (int)
-     * val.getObject(); }
-     *
-     * public static OtpErlangBinary int_to_bin(int val) { return new
-     * OtpErlangBinary(Integer.valueOf(val)); }
+     * Shamelessly based on code from the below link. Altered to be more of a
+     * function than an object method
+     * https://www.avajava.com/tutorials/lessons/how-do-i-perform-a-deep-clone-using
+     * -serializable.html
      */
+
+    try {
+      ByteArrayOutputStream baos = new ByteArrayOutputStream();
+      ObjectOutputStream oos = new ObjectOutputStream(baos);
+      oos.writeObject(this);
+
+      ByteArrayInputStream bais = new ByteArrayInputStream(baos.toByteArray());
+      ObjectInputStream ois = new ObjectInputStream(bais);
+      return (CRDT) ois.readObject();
+    } catch (IOException e) {
+      return null;
+    } catch (ClassNotFoundException e) {
+      return null;
```

Comment:
Are these exceptions supposed to be impossible?  If so, then just make a comment like "should be fatal" and throw whatever you're using for a fatal runtime exception.  It doesn't look like deepclone returning null is handled in the Backend class at all, so it probably shouldn't do so here.

Reply:
Makes sense